### ISSUE TITLE
UniFi Protect Controller Firmware update causing 500 response (unable to retrieve NVR configuration)

### ISSUE THREAD

**Issue by dblommesteijn** [hjdhjd/homebridge-unifi-protect#781]
**Describe The Problem:**
After updating to the latest version of Cloud Key Plus 2.5.11 Unifi protect plugin is unable to connect to my Cloud Key.

**To Reproduce:**
- Update Cloud Key Plus to version 2.5.11
- Restart homebrigde.io (make sure it's getting the NVR configuration information)

**Logs:**
```
[17/08/2022, 10:39:58] [UniFi Protect] API access error: 500 - Internal Server Error
[17/08/2022, 10:39:58] [UniFi Protect] 192.168.1.xxx: Unable to retrieve NVR configuration information from UniFi Protect. Will retry again later.
[17/08/2022, 10:39:59] [UniFi Protect] API access error: 500 - Internal Server Error
[17/08/2022, 10:39:59] [UniFi Protect] 192.168.1.xxx: Unable to retrieve NVR configuration information from UniFi Protect. Will retry again later.
[17/08/2022, 10:39:59] [UniFi Protect] API access error: 500 - Internal Server Error
[17/08/2022, 10:39:59] [UniFi Protect] 192.168.1.xxx: Unable to retrieve NVR configuration information from UniFi Protect. Will retry again later.
[17/08/2022, 10:39:59] [UniFi Protect] API access error: 500 - Internal Server Error
[17/08/2022, 10:39:59] [UniFi Protect] 192.168.1.xxx: Unable to retrieve NVR configuration information from UniFi Protect. Will retry again later.
```

**Homebridge Configuration:**

```json
{
            "controllers": [
                {
                    "address": "192.168.1.xxx",
                    "password": "***",
                    "username": "***"
                }
            ],
            "options": [
                "Disable.Video.HKSV.TimeshiftBuffer"
            ],
            "verboseFfmpeg": true,
            "platform": "UniFi Protect"
        },
```

**Environment:**

* **Homebridge Version:** 1.5.0
* **Node Version:** v16.15.0
* **Homebridge-UniFi-Protect Plugin Version:** v5.5.4
* **UniFi Protect Version:** 2.1.2
* **UniFi Protect Controller Type:** UCKgen2+
* **UniFi Protect Controller Firmware Version:** 2.5.11
* **Apple Device and iOS / macOS / iPadOS / tvOS Version:** _does not apply_
* **Operating System and OS Version:** Raspbian

**Comment by hjdhjd** [comment]
Thanks for the report. The plugin does in fact connect to the Protect controller - the Protect controller is what is producing the errors here. There isn’t anything I can do for you on the plugin end I’m afraid - this is a Protect issue.

You might want to try physically powering off the UCK and then powering it back on, and/or reinstalling the firmware update. Both of these have had good results in the past to varying degrees, but the issue unfortunately lies there. 

I know the above steps seem a bit odd, but give them a shot and see how it goes.

Best of luck,

Habeeb.

**Comment by hjdhjd** [comment]
Thanks for the report. The plugin does in fact connect to the Protect controller - the Protect controller is what is producing the errors here. There isn’t anything I can do for you on the plugin end I’m afraid - this is a Protect issue.

You might want to try physically powering off the UCK and then powering it back on, and/or reinstalling the firmware update. Both of these have had good results in the past to varying degrees, but the issue unfortunately lies there.

I know the above steps seem a bit odd, but give them a shot and see how it goes.

Best of luck!

**Comment by raphus** [comment]
also having many issues after updating the controller version, no more face recognition, constant disconnects:

> `[8/17/2022, 8:32:02 AM] [homebridge-unifi-protect] 192.168.33.7: Unable to retrieve NVR configuration information from UniFi Protect. Will retry again later.
[8/17/2022, 8:32:16 AM] [homebridge-unifi-protect] 192.168.33.7: Controller API connection terminated because it was taking too long. This error can usually be safely ignored.
[8/17/2022, 8:32:16 AM] [homebridge-unifi-protect] 192.168.33.7: Unable to retrieve NVR configuration information from UniFi Protect. Will retry again later.
[8/17/2022, 8:32:30 AM] [homebridge-unifi-protect] 192.168.33.7: Controller API connection terminated because it was taking too long. This error can usually be safely ignored.
[8/17/2022, 8:32:30 AM] [homebridge-unifi-protect] 192.168.33.7: Unable to retrieve NVR configuration information from UniFi Protect. Will retry again later.
[8/17/2022, 8:32:43 AM] [homebridge-unifi-protect] 2204 [UNVR4]: Connected to the UniFi Protect controller API (address: 192.168.33.7 mac: XXXXXXXXXXXXXX).
[8/17/2022, 8:33:09 AM] [homebridge-unifi-protect] 2204 [UNVR4]: Controller API connection terminated because it was taking too long. This error can usually be safely ignored.
[8/17/2022, 8:33:09 AM] [homebridge-unifi-protect] 2204 [UNVR4]: Unable to retrieve NVR configuration information from UniFi Protect. Will retry again later.
[8/17/2022, 8:33:20 AM] [homebridge-unifi-protect] 2204 [UNVR4]: Connected to the UniFi Protect controller API (address: 192.168.33.7 mac: XXXXXXXXXXXXXX).
[8/17/2022, 8:33:45 AM] [homebridge-unifi-protect] 2204 [UNVR4]: Controller API connection terminated because it was taking too long. This error can usually be safely ignored.
[8/17/2022, 8:33:45 AM] [homebridge-unifi-protect] 2204 [UNVR4]: Unable to retrieve NVR configuration information from UniFi Protect. Will retry again later.`

**Comment by hjdhjd** [comment]
@raphus If you have an issue, please open one. Please don't pile onto another user's support request. Thanks!

**Comment by raphus** [comment]
> @raphus If you have an issue, please open one. Please don't pile onto another user's support request. Thanks!
Sincerely apologize, figured it was related, just trying to help. 

@dblommesteijn I was able to get the errors to go away by downgrading the UNVR to a previous version (through SSH), then upgrading again to the latest 2.5.11, no more "Unable to retrieve NVR configuration information from UniFi Protect. Will retry again later." or " Controller API connection terminated because it was taking too long."

**Comment by hjdhjd** [comment]
@raphus Thanks!

As I said above...these issues really reside on the Protect end. Ubiquiti's 2.1.x production releases are still in relatively early days and these issues aren't uncommon unfortunately. They are usually resolved by reinstalling the firmware, downgrading and reupgrading, or a full power cycle. This is especially true at the lower end of the power curve (e.g. UCK).

Thanks again!

**Comment by dblommesteijn** [comment]
@hjdhjd Fair enough, I'm just trying to figure out what the problem is. I guess It's the Unifi OS 2.5.11 not playing nicely with Protect 2.1.2. When I terminate the Protect service, the errors in Homebrew change from 500 to 404, which makes sense. Another think I have noticed, it's the upgrade from Unifi OS 2.4.10 to 2.5.11 is causing the problem, as my logs show after this upgrade the error showed up.

**The steps I took since (without any luck)**
- Power cycle, turn off, disconnect usb
- Try to install firmware for Unify OS 2.4.10 (UCK errors with: `ERROR:  /var/tmp/fwupdate.oPJUv5UIXv (Firmware: UCKP.apq8053.v2.4.10.333a697.220505.2148) doesn't fit the system!`)
- Restored a backups from before 2.5.11 update (however the version is not affected)
- Downgraded Protect to 2.0.1
- Factory reset unit, restored afterwards to 2.5.11 
- Downgraded Protect to 1.12.6

@hjdhjd Anything you can think of besides removing everything from Unifi and start over without a restored session?

**Comment by dblommesteijn** [comment]
**Solved my problem:**

I did a complete console-reset, without recovery from backup.
Configured everything again, connected to Homebridge with this plugin.
Problem solved.

Unifi OS version 2.5.11
Protect version: 2.1.2

_I am shocked it's working again to be honest, but happy it is._

Closing this issue, as it's a Unifi local problem.

**Comment by raphus** [comment]
> **Solved my problem:**
> 
> I did a complete console-reset, without recovery from backup. Configured everything again, connected to Homebridge with this plugin. Problem solved.
> 
> Unifi OS version 2.5.11 Protect version: 2.1.2
> 
> _I am shocked it's working again to be honest, but happy it is._
> 
> Thanks for the quick response, and tips.
> 
> Closing this issue, as it's a Unifi local problem.


Nice, glad to hear you're good. May have to do the same full reset myself.. I was fine yesterday but today I am back to getting errors:
```
[8/18/2022, 10:10:07 AM] [homebridge-unifi-protect] Invalid login credentials given. Please check your login and password.
[8/18/2022, 10:10:07 AM] [homebridge-unifi-protect] 192.168.33.7: API endpoint access error
[8/18/2022, 10:10:07 AM] [homebridge-unifi-protect] 192.168.33.7: Unable to retrieve the livestream websocket API endpoint from the UniFi Protect controller.
[8/18/2022, 10:34:28 AM] [homebridge-unifi-protect] 2204 [UNVR4]: Unable to retrieve NVR configuration information from UniFi Protect.

```
Strangely mine had no errors for about 5 hours after i re-installed firmware, then started again. I noticed "node" process using high CPU (80%-150%), and load averages close to 3 on the NVR. Curious what you are seeing on your end. I have 8 cameras, 7 G4 bullet, 1 G4 pro.

**Comment by hjdhjd** [comment]
> **Solved my problem:**
> 
> I did a complete console-reset, without recovery from backup. Configured everything again, connected to Homebridge with this plugin. Problem solved.
> 
> Unifi OS version 2.5.11 Protect version: 2.1.2
> 
> _I am shocked it's working again to be honest, but happy it is._
> 
> Thanks for the quick response, and tips.
> 
> Closing this issue, as it's a Unifi local problem.

Almost always will be with these updates from Ubiquiti. Everyone is usually pretty quick to update to their latest/greatest, but history and track record show that it's a very perilous path, especially when they are early in a new major codebase update. I would expect more of this in the near-term...historically it takes them a good 6-8 months to stabilize production firmwares after they release the first one. So...sometime in 2023 I'd expect the 2.x codebase to feel more reliable day-to-day for applications and plugins like these to be pretty seamless when updates occur. 

They're not regression-testing against HBUP unfortunately. 😄 

We do happen to make very liberal use of their APIs that they probably don't stress test regularly to the degree that I push the Protect controller...so what Ubiquiti perceives as an "edge" case turns out at times to create very real weirdness for HBUP.

Final note - one thing you can count on is that if there's a genuine issue with the plugin and production firmware releases, I will get it addressed ASAP (as I did when 2.0 came out for instance). I use this every single day of my life...so it's critical code in my household. 😄 

Note to others who might read this: this doesn't mean your particular perceived quirk/feature/dislike-of-something is getting addressed on a critical-fix time horizon. 😉

Archiving this issue.
